spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -0,0 +1,181 @@
+
+/**
+ * An implementation of {@link DatastoreOperations}.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class DatastoreTemplate implements DatastoreOperations {
+
+	private final Datastore datastore;
+
+	private final DatastoreEntityConverter datastoreEntityConverter;
+
+	private final DatastoreMappingContext datastoreMappingContext;
+
+	public DatastoreTemplate(Datastore datastore,
+			DatastoreEntityConverter datastoreEntityConverter,
+			DatastoreMappingContext datastoreMappingContext) {
+		Assert.notNull(datastore, "A non-null Datastore service object is required.");
+		Assert.notNull(datastoreEntityConverter,
+				"A non-null DatastoreEntityConverter is required.");
+		Assert.notNull(datastoreMappingContext,
+				"A non-null DatastoreMappingContext is required.");
+		this.datastore = datastore;
+		this.datastoreEntityConverter = datastoreEntityConverter;
+		this.datastoreMappingContext = datastoreMappingContext;
+	}
+
+	@Override
+	public <T> T findById(Object id, Class<T> entityClass) {
+		return this.datastoreEntityConverter.read(entityClass,
+				this.datastore.get(getKeyFromId(id, entityClass)));
+	}
+
+	@Override
+	public <T> void save(T instance) {
+		this.datastore.put(convertToEntity(instance));
+	}
+
+	@Override
+	public <T> void deleteById(Object id, Class<T> entityClass) {
+		this.datastore.delete(getKeyFromId(id, entityClass));
+	}
+
+	@Override
+	public <T> void delete(T entity) {
+		this.datastore.delete(getKey(entity));
+	}
+
+	@Override
+	public void deleteAll(Class<?> entityClass) {
+		this.datastore.delete(findAll(entityClass).parallelStream().map(this::getKey)
+				.toArray(Key[]::new));
+	}
+
+	@Override
+	public long count(Class<?> entityClass) {
+		return findAll(entityClass).size();
+	}
+
+	@Override
+	public <T> Collection<T> findAllById(Iterable<?> ids, Class<T> entityClass) {
+		List<Key> keysToFind = new ArrayList<>();
+		ids.forEach(x -> keysToFind.add(getKeyFromId(x, entityClass)));
+		return convertEntities(
+				this.datastore.get(keysToFind.toArray(new Key[keysToFind.size()])),
+				entityClass);
+	}
+
+	@Override
+	public <T> Collection<T> findAll(Class<T> entityClass) {
+		return convertEntities(
+				this.datastore.run(Query.newEntityQueryBuilder()
+						.setKind(this.datastoreMappingContext
+								.getPersistentEntity(entityClass).kindName())
+						.build()),
+				entityClass);
+	}
+
+	@Override
+	public <T> boolean existsById(Object id, Class<T> entityClass) {
+		return findById(id, entityClass) != null;
+	}
+
+	@VisibleForTesting
+	Key getKeyFromId(Object id, Class entityClass) {
+		Assert.notNull(id, "Cannot get key for null ID value.");
+		if (id instanceof Key) {
+			return (Key) id;
+		}
+		KeyFactory keyFactory = this.datastore.newKeyFactory();
+		keyFactory.setKind(
+				this.datastoreMappingContext.getPersistentEntity(entityClass).kindName());
+		Key key;
+		if (id instanceof String) {
+			key = keyFactory.newKey((String) id);
+		}
+		else if (id instanceof Long) {
```

Comment:
just to clarify, did you mean long or Long?

Reply:
So the function newKey takes  long, (small l). However you can't do "instance of long". If you do "id.getClass() == long.class", then that will always be false because a long.getClass() is actually Long.